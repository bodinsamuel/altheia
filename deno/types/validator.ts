import { TypeBase } from '../validators/index.ts';

export interface ValidatorOptions {
  required?: boolean;
  unknown?: boolean;
  flatten?: boolean;
}

export interface ValidatorConfirm {
  initial: string;
  comparison: string;
}

export interface ValidatorSchema {
  [k: string]: TypeBase;
}
