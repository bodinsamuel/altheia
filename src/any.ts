import Base from './base';

/**
 * Any is just a empty class to allow chaining
 */
export class TypeAny extends Base {
  /**
   * Constructor
   */
  constructor() {
    super();
    this.name = 'any';
  }
}

const def = {
  Class: TypeAny,
};

export default def;
