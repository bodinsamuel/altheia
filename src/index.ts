import isPlainObject from 'lodash/isPlainObject';

import LangBase from './lang';
import Validator from './validator';
import TypeBase from './base';
import {
  LangList,
  AltheiaInstance,
  ValidatorSchema,
  ValidatorErrorFormatted,
  ValidatorErrorRaw,
} from './types';

// Default extractors
import AnyValidator from './any';
import ArrayValidator from './array';
import BooleanValidator from './boolean';
import DateValidator from './date';
import FunctionValidator from './function';
import InternetValidator from './internet';
import NumberValidator from './number';
import ObjectValidator from './object';
import StringValidator from './string';

export const Instance = (lang?: LangList): AltheiaInstance => {
  // @ts-ignore
  const inst: AltheiaInstance = (schema: ValidatorSchema): Validator => {
    return new Validator(schema, inst);
  };

  inst.langList = Object.assign({}, LangBase);
  inst.templates = {};
  inst.Base = TypeBase;

  inst.instance = (newLang: LangList = {}): AltheiaInstance => {
    return Instance(Object.assign({}, inst.langList, newLang));
  };

  inst.use = (plugin): AltheiaInstance => {
    const test = new plugin.Class();
    inst[test.name || test.constructor.name] = (): TypeBase => {
      return new plugin.Class(inst);
    };
    if (plugin.messages) {
      inst.lang(plugin.messages);
    }
    return inst;
  };

  inst.lang = (key, tpl): void => {
    if (isPlainObject(key)) {
      inst.langList = Object.assign({}, inst.langList, key);
    } else if (typeof key === 'string' && tpl) {
      inst.langList[key] = tpl;
    }
  };

  inst.template = (name, schema): void => {
    inst.templates[name] = schema;
  };

  inst.is = (name): TypeBase => {
    if (typeof inst.templates[name] === 'undefined') {
      throw new Error(`unknow template ${name}`);
    }

    return inst.templates[name].clone();
  };

  inst.formatError = (
    { type, args, result },
    label = 'value',
    position
  ): ValidatorErrorFormatted => {
    // Get messages from error
    let msg;
    if (typeof inst.langList[type] !== 'undefined') {
      msg = inst.langList[type](label, args, result);
    } else {
      msg = 'Invalid value';
    }

    const formatted: ValidatorErrorFormatted = {
      label,
      type,
      message: msg,
    };

    if (typeof position === 'number') {
      formatted.position = position;
    }

    // nested errors
    if (result && result.errors) {
      formatted.errors = result.errors.map((error: ValidatorErrorRaw) =>
        inst.formatError(error.test, error.label, error.position)
      );
    }
    return formatted;
  };

  // Declare basic plugins
  inst.use(AnyValidator);
  inst.use(ArrayValidator);
  inst.use(BooleanValidator);
  inst.use(DateValidator);
  inst.use(FunctionValidator);
  inst.use(InternetValidator);
  inst.use(NumberValidator);
  inst.use(ObjectValidator);
  inst.use(StringValidator);

  // Add passed lang object
  if (lang && isPlainObject(lang)) {
    inst.lang(lang);
  }

  return inst;
};

export default Instance();
