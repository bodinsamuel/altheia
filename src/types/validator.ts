import { TypeBase } from '../validators';

export interface ValidatorOptions {
  required?: boolean;
  unknown?: boolean;
  flatten?: boolean;
}

export interface ValidatorConfirm {
  initial: string;
  comparison: string;
}

export interface ValidatorSchema {
  [k: string]: TypeBase;
}
