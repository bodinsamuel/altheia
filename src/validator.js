const isPlainObject = require('lodash/isPlainObject');
const isEqual = require('lodash/isEqual');

const ObjectValidator = require('./object').Class;

/**
 * Validator class
 * new Validator({ foo: 'bar' });
 */
class Validator {
  /**
   * Constructor
   * @param  {object} schema
   * @param  {object} inst   An Altheia instance
   * @return {Validator}
   */
  constructor(schema, inst) {
    this.inst = inst;
    this.isValidator = 1;
    this._schema = {};
    this._body = {};
    this._errors = [];
    this._errorsRaw = [];
    this._confirm = [];
    this._options = {
      required: false,
      unknown: true,
    };

    this.schema(schema);
  }

  /**
   * Assign the body to validate
   *
   * @param  {object} body
   * @return {Validator}
   */
  body(body) {
    this._body = Object.assign({}, body);
    return this;
  }

  /**
   * Declare the schema that describe the body()
   *
   * @param  {object} schema
   * @return {Validator}
   */
  schema(schema) {
    if (!isPlainObject(schema)) {
      throw new Error('schema should be object');
    }
    this._schema = schema;
    return this;
  }

  /**
   * Declare options to change the defaults behaviour of the Validator
   *
   * @param  {object} options
   * @return {Validator}
   */
  options(options) {
    if (!isPlainObject(options)) {
      throw new Error('schema should be object');
    }
    this._options = Object.assign(this._options, options);
    return this;
  }

  /**
   * Format any Error Array returned by a test
   * @param  {object} error
   * @param  {string} label
   * @return {object}
   */
  formatError(error, label) {
    return this.inst.formatError(error, label);
  }

  /**
   * Validate the body against the schema
   *
   * @param  {Function} callback
   * @return {object}
   */
  async validate(callback = null) {
    // Return an object and call a callback if needed
    const returnOrCallback = (callback, result) => {
      if (callback) {
        callback(result);
      }
      return result;
    };


    const errors = [];
    // Early return if unknown keys
    if (this._options.unknown === false) {
      const only = await new ObjectValidator()
        .in(Object.keys(this._schema), { oneErrorPerKey: true })
        .validate(this._body);
      if (only) {
        only.result.errors.map(error => errors.push(this.formatError(error.test, error.label)));
      }
    }

    // Use old syntax to allow await in loop without using promise.all
    const keys = Object.keys(this._schema);
    for (var i = 0; i < keys.length; i++) {
      const key = keys[i];
      const item = this._schema[key];
      const value =
        typeof this._body[key] !== 'undefined' ? this._body[key] : null;
      this._body[key] = value;

      // If not required pass
      if (this._options.required === true) {
        item.required();
      }

      // validate the item
      const hasError = await item.validate(value);
      if (!hasError) {
        continue;
      }

      this._errorsRaw.push({ test: hasError, label: key });
      const formatted = this.formatError(hasError, key);
      errors.push(formatted);
    }
    this._errors = errors;

    // Check confirm after validation
    if (this._confirm.length > 0) {
      this._confirm.forEach((item) => {
        const initial = this._body[item.initial];
        const comparison = this._body[item.comparison];
        if (isEqual(initial, comparison)) {
          return;
        }
        this._errors.push(
          this.formatError({ name: 'confirm', args: item }, item.comparison)
        );
      });
    }

    if (errors.length > 0) {
      return returnOrCallback(callback, errors);
    }

    return returnOrCallback(callback, false);
  }

  /**
   * Force the confirmation of one field by an other one
   *
   * @param  {string} initial    The first key
   * @param  {string} comparison The second key
   * @return {Validator}
   */
  confirm(initial, comparison) {
    this._confirm.push({ initial, comparison });
    return this;
  }
}

module.exports = Validator;
