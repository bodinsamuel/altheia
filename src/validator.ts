import isPlainObject from 'lodash/isPlainObject';
import isEqual from 'lodash/isEqual';

import { TypeObject } from './object';
import {
  ValidatorConfirm,
  ValidatorOptions,
  ValidatorSchema,
  AltheiaInstance,
  ValidatorErrorRaw,
  ValidatorErrorFormatted,
  ValidatorTestResult,
} from './types';
import { createTest, createTestResult } from './utils/createTest';

export type ValidatorResult = false | ValidatorErrorFormatted[];
export type ValidatorCallback = (errors: ValidatorResult) => void;
/**
 * Validator class
 * new Validator({ foo: 'bar' });
 */
class Validator {
  inst: AltheiaInstance;
  isValidator: number;
  validated: boolean;

  _schema: any;
  _body: any;
  _errors: ValidatorErrorFormatted[];
  _errorsRaw: ValidatorErrorRaw[];
  _confirm: ValidatorConfirm[];
  _options: ValidatorOptions;

  /**
   * Constructor
   * @param  {object} schema
   * @param  {object} inst   An Altheia instance
   */
  constructor(schema: ValidatorSchema, inst: AltheiaInstance) {
    this.inst = inst;

    this.isValidator = 1;
    this.validated = false;

    this._schema = {};
    this._body = {};
    this._errors = [];
    this._errorsRaw = [];
    this._confirm = [];
    this._options = {
      required: false,
      unknown: false,
      flatten: false,
    };

    this.schema(schema);
  }

  /**
   * Clone a validator
   * @return {Validator}
   */
  clone(): Validator {
    const clone = new Validator(this._schema, this.inst);
    clone.options(this._options);
    clone._confirm = [...this._confirm];
    return clone;
  }

  /**
   * Assign the body to validate
   *
   * @param  {object} body
   * @return {this}
   */
  body(body: any): this {
    this._body = Object.assign({}, body);
    return this;
  }

  /**
   * Declare the schema that describe the body()
   *
   * @param  {object} schema
   * @return {this}
   */
  schema(schema: ValidatorSchema): this {
    if (!isPlainObject(schema)) {
      throw new Error('schema should be object');
    }
    this._schema = schema;
    return this;
  }

  /**
   * Declare options to change the defaults behaviour of the Validator
   *
   * @param  {object} options
   * @return {this}
   */
  options(options: ValidatorOptions): this {
    if (!isPlainObject(options)) {
      throw new Error('schema should be object');
    }
    this._options = Object.assign(this._options, options);
    return this;
  }

  /**
   * Format any Error Array returned by a test
   * @param  {object} error
   * @param  {string} label
   * @return {ValidatorErrorFormatted} Formatted error
   */
  formatError(
    error: ValidatorTestResult,
    label: string
  ): ValidatorErrorFormatted {
    return this.inst.formatError(error, label);
  }

  private flatten(bag: any[], error: ValidatorErrorFormatted): any[] {
    if (error.errors) {
      return error.errors.reduce(this.flatten, bag);
    }
    bag.push(error);
    return bag;
  }

  /**
   * Validate the body against the schema
   *
   * @param  {function} callback
   * @return {false | ValidatorErrorFormatted[]} Test resutls
   */
  async validate(
    body: any,
    callback?: ValidatorCallback
  ): Promise<ValidatorResult>;
  async validate(callback?: ValidatorCallback): Promise<ValidatorResult>;
  async validate(...params: any | ValidatorCallback): Promise<ValidatorResult> {
    if (this.validated) {
      throw new Error(
        'Already validated, please use .clone() to validate a different body'
      );
    }

    let callback: ValidatorCallback | undefined = undefined;
    if (params.length > 0) {
      if (typeof params[0] === 'function') {
        callback = params[0];
      } else {
        this.body(params[0]);
        callback = params[1];
      }
    }

    this.validated = true;

    // Return an object and call a callback if needed
    const returnOrCallback = (
      result: any,
      callback?: (value: ValidatorResult) => void
    ): ValidatorResult => {
      if (callback) {
        callback(result);
      }
      return result;
    };

    const errors = [];
    // Early return if unknown keys
    if (this._options.unknown === false) {
      const only = await new TypeObject()
        .in(Object.keys(this._schema), { oneErrorPerKey: true })
        .validate(this._body);

      if (typeof only !== 'boolean' && only.result && only.result.errors) {
        only.result.errors.map((error: ValidatorErrorRaw) =>
          errors.push(this.formatError(error.test, error.label))
        );
      }
    }

    // Use old syntax to allow await in loop without using promise.all
    const keys = Object.keys(this._schema);
    for (var i = 0; i < keys.length; i++) {
      const key = keys[i];
      const item = this._schema[key];
      const value =
        typeof this._body[key] !== 'undefined' ? this._body[key] : null;
      this._body[key] = value;

      // If not required pass
      if (this._options.required === true) {
        item.required();
      }

      // validate the item
      const hasError = await item.validate(value);
      if (!hasError) {
        continue;
      }

      this._errorsRaw.push({ test: hasError, label: key });
      const formatted = this.formatError(hasError, key);
      if (!formatted.errors || !this._options.flatten) {
        errors.push(formatted);
      } else if (this._options.flatten && formatted.errors) {
        errors.push(...formatted.errors.reduce(this.flatten, []));
      }
    }
    this._errors = errors;

    // Check confirm after validation
    if (this._confirm.length > 0) {
      this._confirm.forEach((item) => {
        const initial = this._body[item.initial];
        const comparison = this._body[item.comparison];
        if (isEqual(initial, comparison)) {
          return;
        }

        this._errors.push(
          this.formatError(
            createTestResult(
              createTest({ type: 'confirm', args: item }),
              false
            ),
            item.comparison
          )
        );
      });
    }

    if (errors.length > 0) {
      return returnOrCallback(errors, callback);
    }

    return returnOrCallback(false, callback);
  }

  /**
   * Force the confirmation of one field by an other one
   *
   * @param  {string} initial    The first key
   * @param  {string} comparison The second key
   * @return {this}
   */
  confirm(initial: string, comparison: string): this {
    this._confirm.push({ initial, comparison });
    return this;
  }
}

export default Validator;
