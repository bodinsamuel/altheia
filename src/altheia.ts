import isPlainObject from 'lodash/isPlainObject';

import LangBase from './lang';
import { Validator } from './validator';
import { TypeBase } from './validators/base';
import { LangList } from './types/lang';
import { AltheiaInstance } from './types/instance';
import { ValidatorErrorFormatted, ValidatorErrorRaw } from './types/tests';
import { ValidatorSchema } from './types/validator';

// Default extractors
import AnyValidator from './validators/any';
import ArrayValidator from './validators/array';
import BooleanValidator from './validators/boolean';
import DateValidator from './validators/date';
import FunctionValidator from './validators/function';
import InternetValidator from './validators/internet';
import NumberValidator from './validators/number';
import ObjectValidator from './validators/object';
import StringValidator from './validators/string';

const Instance = (lang?: LangList): AltheiaInstance => {
  const inst: AltheiaInstance = ((schema: ValidatorSchema): Validator => {
    return new Validator(schema, inst);
  }) as AltheiaInstance;

  inst.langList = { ...LangBase };
  inst.templates = {};
  inst.Base = TypeBase;

  inst.instance = (newLang: LangList = {}): AltheiaInstance => {
    return Instance({ ...inst.langList, ...newLang });
  };

  inst.use = (plugin): AltheiaInstance => {
    const test = new plugin.Class();
    inst[test.name || test.constructor.name] = (): TypeBase => {
      return new plugin.Class(inst);
    };
    if (plugin.messages) {
      inst.lang(plugin.messages);
    }
    return inst;
  };

  inst.lang = (key, tpl): void => {
    if (typeof key === 'object' && isPlainObject(key)) {
      inst.langList = { ...inst.langList, ...key };
    } else if (typeof key === 'string' && tpl) {
      inst.langList[key] = tpl;
    }
  };

  inst.template = (name, schema): void => {
    inst.templates[name] = schema;
  };

  inst.is = (name): TypeBase => {
    if (typeof inst.templates[name] === 'undefined') {
      throw new Error(`unknow template ${name}`);
    }

    return inst.templates[name].clone();
  };

  inst.formatError = (
    { type, args, result },
    label = 'value',
    position
  ): ValidatorErrorFormatted => {
    // Get messages from error
    let msg;
    if (typeof inst.langList[type] !== 'undefined') {
      msg = inst.langList[type](label, args, result);
    } else {
      msg = 'Invalid value';
    }

    const formatted: ValidatorErrorFormatted = {
      label,
      type,
      message: msg,
    };

    if (typeof position === 'number') {
      formatted.position = position;
    }

    // nested errors
    if (result && result.errors) {
      formatted.errors = result.errors.map(
        (error: ValidatorErrorRaw): ValidatorErrorFormatted =>
          inst.formatError(error.test, error.label, error.position)
      );
    }
    return formatted;
  };

  // Declare basic plugins
  inst.use(AnyValidator);
  inst.use(ArrayValidator);
  inst.use(BooleanValidator);
  inst.use(DateValidator);
  inst.use(FunctionValidator);
  inst.use(InternetValidator);
  inst.use(NumberValidator);
  inst.use(ObjectValidator);
  inst.use(StringValidator);

  // Add passed lang object
  if (lang && isPlainObject(lang)) {
    inst.lang(lang);
  }

  return inst;
};

export const Altheia = Instance();
