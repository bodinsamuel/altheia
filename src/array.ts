import TypeBase from './base';
import arrayDiff from './utils/arraydiff';
import { LangList, ValidatorErrorRaw } from './types';

export const messages: LangList = {
  'array.typeof': (name) => `${name} must be a valid array`,
  'array.min': (name, args: { min: number }) =>
    `${name} must contains at least ${args.min} items`,
  'array.max': (name, args: { max: number }) =>
    `${name} must contains at most ${args.max} items`,
  'array.in': (name, args: { in: string[] }) =>
    `${name} must only contains these keys [${args.in}]`,
  'array.not': (name) => `${name} contains forbidden value`,
  'array.unique': (name) => `${name} can not contains duplicated value`,
  'array.oneOf': (name) => `${name} contains forbidden items`,
  'array.itemInvalid': (name) =>
    `${name} does not match any of the allowed types`,
};

/**
 * Array class
 */
export class TypeArray extends TypeBase {
  /**
   * Constructor
   */
  constructor() {
    super();
    this.name = 'array';
    this.typeof();
  }

  /**
   * Test to validate the type of the value
   *
   * @return {this}
   */
  typeof(): this {
    this.test('typeof', (val: any) => {
      return Array.isArray(val);
    });
    return this;
  }

  /**
   * Force an array to contains at least a number of items equal to the value passed.
   *
   * @param  {number} min
   * @return {this}
   */
  min(min: number): this {
    this.test(
      'min',
      (arr: any[]) => {
        return arr.length >= min;
      },
      { min }
    );

    return this;
  }

  /**
   * Force an array to contains at most a number of items equal to the value passed.
   *
   * @param  {number} max
   * @return {this}
   */
  max(max: number): this {
    this.test(
      'max',
      (arr: any[]) => {
        return arr.length <= max;
      },
      { max }
    );

    return this;
  }

  /**
   * Force an array to have only the values passed in the set
   *
   * @param  {...array} array
   * @return {this}
   */
  in(value: any[]): this;
  in(...value: any): this;
  in(...array: any): this {
    let only = array;
    // handle someone passing literal array instead of multiple args
    if (array.length === 1 && Array.isArray(array[0])) {
      only = (array[0] as unknown) as any[];
    }

    this.test(
      'in',
      (arr: any[]) => {
        return arrayDiff<string>(arr, only).length === 0;
      },
      { in: only }
    );

    return this;
  }

  /**
   * Force an array not to have the values passed in the set
   *
   * @param  {...array} array
   * @return {this}
   */
  not(value: any[]): this;
  not(...value: any): this;
  not(...array: any[]): this {
    let only = array;
    // handle someone passing literal array instead of multiple args
    if (array.length === 1 && Array.isArray(array[0])) {
      only = (array[0] as unknown) as any[];
    }

    this.test(
      'not',
      (arr: any[]) => {
        return arrayDiff<string>(only, arr).length === only.length;
      },
      { not: only }
    );

    return this;
  }

  /**
   * Force an array to only have each item once
   *
   * @return {this}
   */
  unique(): this {
    this.test('unique', (arr: any[]) => {
      const a = new Set(arr);
      return a.size === arr.length;
    });

    return this;
  }

  /**
   * Force all array's items to match one of the template
   *
   * @param  {...Validator} templates
   * @return {this}
   */
  oneOf(...templates: TypeBase[]): this {
    this.test(
      'oneOf',
      async (arr: any[]) => {
        const errors: ValidatorErrorRaw[] = [];

        await Promise.all(
          arr.map(async (value: any, index: number) => {
            let error: ValidatorErrorRaw | undefined = undefined;

            const label = 'item';
            for (var i = 0; i < templates.length; i++) {
              const test = await templates[i].required().validate(value);
              if (test) {
                error = { label, test, position: index };
              } else {
                // early break if one template matched (returned no error)
                return;
              }
            }

            // Help typescript understand we have error here
            if (!error) {
              return;
            }

            // if multiples templates, return a generic message
            if (templates.length > 1) {
              error = {
                label,
                test: this.createTestResult(
                  this.createTest({
                    type: 'array.itemInvalid',
                  }),
                  false
                ),
                position: index,
              };
              errors.push(error);
            } else {
              errors.push(error);
            }
          })
        );

        return {
          valid: errors.length <= 0,
          error: 'not_matching',
          errors,
        };
      },
      { templates }
    );

    return this;
  }
}

const def = {
  Class: TypeArray,
  messages,
};

export default def;
