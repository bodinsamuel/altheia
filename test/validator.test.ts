import Alt from '../src';
import { Validator } from '../src/validator';

describe('Validator', () => {
  test('Should be a func', async () => {
    expect(Alt).toBeInstanceOf(Function);
  });

  test('Should validate with callback', async () => {
    let mark = false;
    await Alt({ foo: Alt.number() })
      .body({ foo: 1 })
      .validate((error) => {
        expect(error).toBe(false);
        mark = true;
      });

    // to avoid false positive we check the callback real use
    expect(mark).toBe(true);
  });

  test('Should validate with callback and error', async () => {
    let mark = false;
    await Alt({ foo: Alt.string() })
      .body({ foo: 1 })
      .validate((error) => {
        expect(error).toBeTruthy();
        mark = true;
      });

    // to avoid false positive we check the callback real use
    expect(mark).toBe(true);
  });

  test('Should validate with await', async () => {
    const result = await Alt.string().validate('foobar');
    expect(result).toBe(false);
  });

  test('Should validate with error', async () => {
    const result = await Alt({ foo: Alt.string() })
      .body({ foo: 1 })
      .validate();
    expect(result).toBeTruthy();
  });

  test('Should validate with cast and not modify primitive', async () => {
    const data = { foo: '1' };
    const result = await Alt({ foo: Alt.number().cast() })
      .body(data)
      .validate();
    expect(result).toBe(false);
    expect(data).toEqual({ foo: '1' });
  });

  test('Should validate with error non i18n', async () => {
    const hasError = await Alt({ foo: Alt.string().custom('my', () => false) })
      .body({ foo: '1' })
      .validate();
    expect(hasError).toBeTruthy();
    expect(hasError).toEqual([
      {
        label: 'foo',
        type: 'string.custom.my',
        message: 'Invalid value',
      },
    ]);
  });
  describe('schema()', () => {
    test('should pass', async () => {
      const instance = Alt({ foo: Alt.string() });
      expect(instance).toBeInstanceOf(Validator);
    });
    test('should throw error on bad params', async () => {
      expect(() => {
        // @ts-ignore
        Alt(true);
      }).toThrow('schema should be object');
    });
  });

  describe('options()', () => {
    test('should throw error on bad params', async () => {
      expect(() => {
        // @ts-ignore
        Alt({}).options(true);
      }).toThrow('schema should be object');
    });

    describe('required', () => {
      test('should fail', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
          password: Alt.string(),
        })
          .body({
            name: 'top',
          })
          .options({
            required: true,
          })
          .validate();

        expect(hasError).toBeTruthy();
        expect(hasError).toEqual([
          {
            label: 'login',
            message: 'login is required',
            type: 'required',
          },
          {
            label: 'password',
            message: 'password is required',
            type: 'required',
          },
        ]);
      });

      test('should pass', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
        })
          .body({
            name: 'top',
          })
          .options({
            required: false,
          })
          .validate();

        expect(hasError).toBe(false);
      });

      test('should fail with empty object', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
        })
          .body({})
          .options({
            required: true,
          })
          .validate();

        expect(hasError).toBeTruthy();
      });

      test('should fail with empty string', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
        })
          .body({
            name: '',
            login: '',
          })
          .options({
            required: true,
          })
          .validate();

        expect(hasError).toBeTruthy();
      });

      test('should fail with empty string with flag but not required', async () => {
        const hasError = await Alt({
          name: Alt.string().noEmpty(),
          login: Alt.string().noEmpty(),
        })
          .body({
            name: '',
            login: '',
          })
          .options({
            required: false,
          })
          .validate();

        expect(hasError).toBeTruthy();
      });

      test('should fail with empty primitive', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
        })
          .body({
            name: null,
            login: false,
          })
          .options({
            required: true,
          })
          .validate();

        expect(hasError).toBeTruthy();
      });
    });

    describe('unknown', () => {
      test('should fail', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
        })
          .body({
            name: 'top',
            foo: 'bar',
          })
          .options({
            unknown: false,
          })
          .validate();

        expect(hasError).toBeTruthy();
        expect(hasError).toEqual([
          {
            label: 'foo',
            message: 'foo is not allowed',
            type: 'forbidden',
          },
        ]);
      });

      test('should fail with different errors', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
        })
          .body({
            name: 'top',
            login: 1,
            foo: 'bar',
          })
          .options({
            unknown: false,
          })
          .validate();

        expect(hasError).toBeTruthy();
        expect(hasError).toEqual([
          {
            label: 'foo',
            message: 'foo is not allowed',
            type: 'forbidden',
          },
          {
            label: 'login',
            message: 'login must be a valid string',
            type: 'string.typeof',
          },
        ]);
      });

      test('should pass', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
        })
          .body({
            name: 'top',
            login: 'bar',
          })
          .options({
            unknown: false,
          })
          .validate();

        expect(hasError).toBe(false);
      });

      test('should pass with empty', async () => {
        const hasError = await Alt({
          name: Alt.string(),
          login: Alt.string(),
        })
          .body({})
          .options({
            unknown: false,
            required: false,
          })
          .validate();

        expect(hasError).toBe(false);
      });
    });

    describe('flatten', () => {
      test('should flatten simple errors', async () => {
        const hasError = await Alt({
          name: Alt.number(),
          login: Alt.number(),
        })
          .body({
            name: 'top',
            login: 'bar',
          })
          .options({
            flatten: true,
          })
          .validate();

        expect(hasError).toMatchSnapshot();
      });

      test('should flatten nested errors', async () => {
        const hasError = await Alt({
          name: Alt.array().oneOf(Alt.number()),
          login: Alt.number(),
        })
          .body({
            name: ['top', 'foo'],
            login: 'bar',
          })
          .options({
            flatten: true,
          })
          .validate();

        expect(hasError).toMatchSnapshot();
      });
    });
  });

  describe('validate()', () => {
    test('Should validate simple schema', async () => {
      const hasError = await Alt({
        login: Alt.string(),
      })
        .body({ login: 'foobar' })
        .validate();
      expect(hasError).toBe(false);
    });
    test('Should not pass simple schema', async () => {
      const hasError = await Alt({
        login: Alt.string(),
      })
        .body({ login: 1 })
        .validate();
      expect(hasError).toBeTruthy();
    });

    test('Should validate with a body', async () => {
      const hasError = await Alt({
        login: Alt.string().required(),
      }).validate({ login: 'foobar' });
      expect(hasError).toBe(false);
    });

    test('Should validate with a body and a callback', async (done) => {
      await Alt({
        login: Alt.string().required(),
      }).validate({ login: 'foobar' }, (errors) => {
        expect(errors).toBe(false);
        done();
      });
    });

    test('Should pass mid-complexe schema', async () => {
      const hasError = await Alt({
        login: Alt.string()
          .min(6)
          .max(15)
          .pattern(/[a-z]+/),
        password: Alt.string().min(10),
      })
        .body({ login: 'foobar', password: 'foobarfoobar' })
        .validate();
      expect(hasError).toBe(false);
    });

    test('Should not pass mid-complexe schema', async () => {
      const hasError = await Alt({
        login: Alt.string()
          .min(6)
          .max(15)
          .pattern(/^[a-z]+$/),
        password: Alt.string().min(10),
      })
        .body({ login: 'foobA', password: 'foob' })
        .validate();
      expect(hasError).toBeTruthy();
      // @ts-ignore
      expect(hasError.length).toBe(2);

      expect(hasError[0]).toEqual({
        label: 'login',
        type: 'string.min',
        message: 'login must be at least 6 characters long',
      });
      expect(hasError[1]).toEqual({
        label: 'password',
        type: 'string.min',
        message: 'password must be at least 10 characters long',
      });
    });
  });

  describe('custom()', () => {
    test('Should use custom message', async () => {
      Alt.lang('string.custom.sync', () => 'foobar');

      const hasError = await Alt({
        login: Alt.string().custom('sync', (value) => {
          return value === 'barfoor';
        }),
      })
        .body({ login: 'foobar' })
        .validate();

      expect(hasError).toBeTruthy();
      expect(hasError[0].message).toBe('foobar');
    });

    test('Should validate sync', async () => {
      const hasError = await Alt({
        login: Alt.string().custom('sync', (value) => {
          return value === 'foobar';
        }),
      })
        .body({ login: 'foobar' })
        .validate();
      expect(hasError).toBe(false);
    });

    test('Should not validate sync', async () => {
      const hasError = await Alt({
        login: Alt.string().custom('sync', (value) => {
          return value === 'barfoor';
        }),
      })
        .body({ login: 'foobar' })
        .validate();
      expect(hasError).toBeTruthy();
    });

    test('Should validate async', async () => {
      let mark = false;
      const hasError = await Alt({
        login: Alt.string().custom('sync', () => {
          return new Promise((resolve) => {
            setTimeout(() => {
              mark = true;
              resolve(true);
            }, 500);
          });
        }),
      })
        .body({ login: 'foobar' })
        .validate();

      expect(hasError).toBe(false);
      expect(mark).toBe(true);
    });

    test('Should not validate async', async () => {
      let mark = false;
      const hasError = await Alt({
        login: Alt.string().custom('sync', () => {
          return new Promise((resolve, reject) => {
            setTimeout(() => {
              mark = true;
              reject();
            }, 500);
          });
        }),
      })
        .body({ login: 'foobar' })
        .validate();

      expect(hasError).toBeTruthy();
      expect(mark).toBe(true);
    });
  });

  describe('confirm()', () => {
    test('Should confirm correctly', async () => {
      const hasError = await Alt({
        login: Alt.string(),
        relogin: Alt.string(),
      })
        .body({ login: 'foobar', relogin: 'foobar' })
        .confirm('login', 'relogin')
        .validate();
      expect(hasError).toBe(false);
    });
    test('Should fail on confirm', async () => {
      const hasError = await Alt({
        login: Alt.string(),
        relogin: Alt.string(),
      })
        .body({ login: 'foobar', relogin: 'barfoo' })
        .confirm('login', 'relogin')
        .validate();

      expect(hasError).toBeTruthy();
      expect(hasError).toEqual([
        {
          label: 'relogin',
          type: 'confirm',
          message: 'relogin must be the same as login',
        },
      ]);
    });
  });
});

describe('Reuse', () => {
  test('should not be reusable', async (done) => {
    const schema = Alt({
      foo: Alt.string(),
    });

    await schema.body({ foo: 1 }).validate();
    expect(schema._errors).toHaveLength(1);
    expect(schema._errorsRaw).toHaveLength(1);

    try {
      await schema.body({ foo: 'bar' }).validate();
    } catch (e) {
      expect(e.message).toBe(
        'Already validated, please use .clone() to validate a different body'
      );
      done();
    }
  });
});
