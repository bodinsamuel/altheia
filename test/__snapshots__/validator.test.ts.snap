// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator options() flatten should flatten nested errors 1`] = `
Array [
  Object {
    "label": "item",
    "message": "item must be a valid number",
    "position": 0,
    "type": "number.typeof",
  },
  Object {
    "label": "item",
    "message": "item must be a valid number",
    "position": 1,
    "type": "number.typeof",
  },
  Object {
    "label": "login",
    "message": "login must be a valid number",
    "type": "number.typeof",
  },
]
`;

exports[`Validator options() flatten should flatten simple errors 1`] = `
Array [
  Object {
    "label": "name",
    "message": "name must be a valid number",
    "type": "number.typeof",
  },
  Object {
    "label": "login",
    "message": "login must be a valid number",
    "type": "number.typeof",
  },
]
`;
